vpr/src/place/initial_placement.cpp

```diff
@@ -84,36 +151,62 @@ static bool try_random_placement(t_pl_macro pl_macro, PartitionRegion& pr, t_log
+ */
+static bool try_dense_placement(t_pl_macro pl_macro, PartitionRegion& pr, t_logical_block_type_ptr block_type, enum e_pad_loc_type pad_loc_type);
+
 void print_sorted_blocks(const std::vector<ClusterBlockId>& sorted_blocks, const vtr::vector<ClusterBlockId, t_block_score>& block_scores);
 
-static void place_all_blocks(const std::vector<ClusterBlockId>& sorted_blocks,
-                             enum e_pad_loc_type pad_loc_type);
+static int place_all_blocks(enum e_pad_loc_type pad_loc_type);
 
 /**
- * @brief If any blocks are unplaced after initial placement, this routine
- * prints an error message showing the names, types, and IDs of the unplaced blocks
+ * @brief If any blocks are unplaced after each initial placement iteration, this routine
+ * prints an log message showing the names, types, and IDs of the unplaced blocks.
+ *  
+ *  @param init_placement_iter initial placement iteration that finished.
+ *  @param unplaced_blocks Number of failed blocks after init_placement_iter.
  */
-static void print_unplaced_blocks();
+static void print_unplaced_blocks(int init_placement_iter, int unplaced_blocks);
 
-static void print_unplaced_blocks() {
+/**
+ * @brief If any blocks remain unplaced after all initial placement iterations, this routine
+ * throw an error indicating that initial placement can not be done with the current device size or
+ * floorplanning constraints. 
+ *  
+ *  @param unplaced_blocks Number of failed blocks after all placement iterations.
+ */
+static void check_initial_placement_legality(int unplaced_blocks);
+
+static void check_initial_placement_legality(int unplaced_blocks) {
+    if (unplaced_blocks > 0) {
+        VPR_FATAL_ERROR(VPR_ERROR_PLACE,
+                        "%d blocks could not be placed during initial placement, no spaces were available for them on the grid.\n"
+                        "If VPR was run with floorplan constraints, the constraints may be too tight.\n",
+                        unplaced_blocks);
+    }
+}
+
+static void print_unplaced_blocks(int init_placement_iter, int unplaced_blocks) {
     auto& cluster_ctx = g_vpr_ctx.clustering();
     auto& place_ctx = g_vpr_ctx.placement();
 
-    int unplaced_blocks = 0;
+    VTR_LOG("Initial placement iteration %d has finished with %d unplaced blocks\n", init_placement_iter, unplaced_blocks);
 
     for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
         if (place_ctx.block_locs[blk_id].loc.x == INVALID_X) {
-            VTR_LOG("Block %s (# %d) of type %s could not be placed during initial placement\n", cluster_ctx.clb_nlist.block_name(blk_id).c_str(), blk_id, cluster_ctx.clb_nlist.block_type(blk_id)->name);
-            unplaced_blocks++;
+            VTR_LOG("Block %s (# %d) of type %s could not be placed during initial placement iteration %d\n", cluster_ctx.clb_nlist.block_name(blk_id).c_str(), blk_id, cluster_ctx.clb_nlist.block_type(blk_id)->name, init_placement_iter);
```

Comment:
Suggest printing this if VERBOSE >= 1 or this is the last placement attempt (we've failed).